### PR TITLE
Hybrid Quantum-Classical Style Transfer - Paper Implementation

### PR DESCRIPTION
**Pull request by Deeksha-Shanmugam**
This project implements a quantum-enhanced neural style transfer using PyTorch and the Classiq quantum framework. It extracts image features using a pretrained VGG-19 model and transforms them via a quantum circuit to compute a quantum-based similarity loss. The classical style transfer is then augmented with this quantum loss for richer stylization. The output is a stylized image combining content and artistic style.

### CODE REVIEW

File: community/paper_implementation_project/hybrid_style_transfer/Hybrid_Quantum_classical_style_transfer.ipynb [hunk 1]
```diff
@@ -0,0 +1 @@
...el at: https://short.classiq.io/join-slack or open a support ticket at: https://classiq-community.freshdesk.com/support/tickets/new\nQuantum circuit execution error: Argument '1.0581386' to parameter 'theta' of function 'RX' has incompatible type; expected CReal\nIf you need further assistance, please reach out on our Community Slack channel at: https://short.classiq.io/join-slack or open a support ticket at: https://classiq-community.freshdesk.com/support/tickets/new\nQuantum circuit execution error: Argument '0.0' to parameter 'theta' of function 'RX' has incompatible type; expected CReal\nIf you need further assistance, please reach out on our Community Slack channel at: https://short.classiq.io/join-slack or open a support ticket at: https://classiq-community.freshdesk.com/support/tickets/new\nQuantum circuit execution error: Argument '1.0581386' to parameter 'theta' of function 'RX' has incompatible type; expected CReal\nIf you need further assistance, please reach out on our Community Slack channel at: https://short.classiq.io/join-slack or open a support ticket at: https://classiq-community.freshdesk.com/support/tickets/new\n","output_type":"stream"}],"execution_count":10}]}
```
Comment (TomerGoldfriend) [inline community/paper_implementation_project/hybrid_style_transfer/Hybrid_Quantum_classical_style_transfer.ipynb:1]: <p>Thank you&nbsp;﻿<span class="mention" data-value="Deeksha-Shanmugam" data-denotation-char="@">﻿<span contenteditable="false">@Deeksha-Shanmugam</span>﻿</span>  for this PR. Please provide details about the code so we can review it properly:</p><ul><li>Short introduction, in particular, you shall specify the paper you are implementing including a reference.</li><li>Explain major code blocks, what is implemented and how it is related to the algorithm.</li><li>This looks like a very complex hybrid neural network, please provide its diagram as a figure. I think it will make the notebook clearer.</li><li>Add a snapshot of the quantum part from Classiq visualization.</li></ul><p>You can have a look at several examples in our library. It should not be identical in terms of sectioning and wording, but just to understand the rational. Here are some QML examples:&nbsp;<a href="https://github.com/Classiq/classiq-library/blob/main/algorithms/qml/hybrid_qnn/hybrid_qnn_for_subset_majority.ipynb" target="_blank">https://github.com/Classiq/classiq-library/blob/main/algorithms/qml/hybrid_qnn/hybrid_qnn_for_subset_majority.ipynb</a>,&nbsp;<a href="https://github.com/Classiq/classiq-library/blob/main/algorithms/qml/quantum_autoencoder/quantum_autoencoder.ipynb" target="_blank">https://github.com/Classiq/classiq-library/blob/main/algorithms/qml/quantum_autoencoder/quantum_autoencoder.ipynb</a></p><p><a href="https://github.com/Classiq/classiq-library/blob/main/algorithms/qml/qgan/qgan_bars_and_strips.ipynb" target="_blank">https://github.com/Classiq/classiq-library/blob/main/algorithms/qml/qgan/qgan_bars_and_strips.ipynb</a></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/970/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (Deeksha-Shanmugam) [reply community/paper_implementation_project/hybrid_style_transfer/Hybrid_Quantum_classical_style_transfer.ipynb:1]: Hi @TomerGoldfriend , 
Thank you for your feedback and for giving me the opportunity to clarify. Here's a full explanation of the project for your review:

Introduction
This project implements a Hybrid Quantum-Classical Neural Style Transfer method, inspired by the paper ["Quantum Style Transfer"](https://arxiv.org/abs/2212.10431).
The goal is to enhance classical artistic style transfer by incorporating quantum computing for feature transformation and optimization.

We combine:

- Classical deep learning (VGG19 feature extraction)
- Quantum circuits (for feature transformation and fidelity-based optimization)
- A new Quantum Fidelity Loss alongside classical content and style losses.

Explanation of Major Code Blocks:

1. Preprocessing and Feature Extraction

- Load content and style images.
- Preprocess images using torchvision transforms.
- Extract multi-layered feature maps using a pretrained VGG-19 network.

2. Quantum Feature Transformation

- Classical features from a selected VGG layer (conv4_1) are flattened.
- Using Classiq’s CircuitBuilder, a quantum circuit is designed:
  - Apply Hadamard gates.
  - Apply parameterized RX rotations based on extracted classical feature values.
  - Simulate the resulting quantum statevector.

3. Quantum Fidelity Loss

- Compute an approximate quantum fidelity between classical and quantum-transformed features.
- Define a Quantum Fidelity Loss that measures 1 - fidelity.
- Add this to the total loss during training to guide optimization.

4. Hybrid Training Loop

- Optimize a generated image by minimizing the weighted sum of:
- Content loss (classical)
- Style loss (classical)
- Quantum fidelity loss (quantum)
- Update the generated image iteratively.

5. Output

- Save the final stylized image after training.

I have attached the architecture and circuit visualization for your reference.
image
image

Comment (TomerGoldfriend) [reply community/paper_implementation_project/hybrid_style_transfer/Hybrid_Quantum_classical_style_transfer.ipynb:1]: Thank you for the clarification @Deeksha-Shanmugam. I still do not fully understand the goal.
Is the original paper you cited uses quantum computing? I am not sure.
So you are taking a classical paper in Machine Learning and replace one part of it by a quantum counterpart?

Comment (Deeksha-Shanmugam) [reply community/paper_implementation_project/hybrid_style_transfer/Hybrid_Quantum_classical_style_transfer.ipynb:1]: Hi @TomerGoldfriend,

Thank you again for your feedback! I appreciate the opportunity to clarify further.

Regarding your questions:
The reference paper I initially mentioned ("QuantArt") does not use quantum computing.
It introduces vector quantization (a classical machine learning technique) to improve visual fidelity in style transfer.
The "Quant" in "QuantArt" refers to quantization, not quantum computing.

About my project:

My project is titled Hybrid Quantum-Classical Neural Style Transfer, and the name reflects the goal accurately.

It is not just a copy of a classical paper — it extends traditional neural style transfer by introducing quantum computing concepts into the process.

Specifically:

Instead of using only classical latent space manipulation (like in QuantArt),

I encode classical features into quantum states using a quantum circuit (with feature maps and variational quantum layers).

I also introduce a Quantum Fidelity Loss, computed by comparing the quantum-encoded feature states, to guide optimization along with classical content and style losses.

Thus, it’s a true hybrid pipeline:

Classical side: VGG feature extraction, preprocessing, postprocessing.

Quantum side: Feature transformation and additional loss computation using quantum circuits.

Comment (TomerGoldfriend) [reply community/paper_implementation_project/hybrid_style_transfer/Hybrid_Quantum_classical_style_transfer.ipynb:1]: <p><span class="mention" data-value="Deeksha-Shanmugam" data-denotation-char="@">﻿<span contenteditable="false">@Deeksha-Shanmugam</span>﻿</span> thank you for the clarification.</p><p>First, please go over the <a href="https://github.com/Classiq/classiq-library/blob/main/CONTRIBUTING.md" target="_blank">Contribution Guidelines</a> and follow the instructions there. In particular, please note that your notebook should run end-to-end, it does not seem like the case here.</p><p>Where do I see training, inference etc in your code? this is supposed to be a hybrid machine learning application. Please go over our tutorials and notebooks for integrating Classiq with PyTorch. For example, you have <a href="https://github.com/Classiq/classiq-library/blob/main/tutorials/basic_tutorials/qml_with_classiq_guide/qml_with_classiq_guide.ipynb" target="_blank">https://github.com/Classiq/classiq-library/blob/main/tutorials/basic_tutorials/qml_with_classiq_guide/qml_with_classiq_guide.ipynb</a>, and <a href="https://github.com/Classiq/classiq-library/blob/main/algorithms/qml/hybrid_qnn/hybrid_qnn_for_subset_majority.ipynb" target="_blank">https://github.com/Classiq/classiq-library/blob/main/algorithms/qml/hybrid_qnn/hybrid_qnn_for_subset_majority.ipynb</a></p>

Comment (TomerGoldfriend) [reply community/paper_implementation_project/hybrid_style_transfer/Hybrid_Quantum_classical_style_transfer.ipynb:1]: @Deeksha-Shanmugam are you planning to work on this and address the general comments above?

**Comment by TomerGoldfriend** [comment]
@Deeksha-Shanmugam I am closing this, you can reopen after addressing the general comments above, in case you still want to work on this.
